### ISSUE TITLE
jsloader fails on XHTML pages

### ISSUE THREAD

**Issue by ForbiddenEra** [google/closure-library#986]
EDIT: I believe I have discovered the root of this issue, I have placed a TL;DR in my ~~last~~ this https://github.com/google/closure-library/issues/986#issuecomment-527825263 post to save anyone the headache of reading through my debugging if they don't need/want to.

Hi,

In attempting to use Firebase for Google/FB/etc authentication, I ran into an issue where the popup authorization method wasn't working. The error seems to have been traced down to firebase depending on closure for loading external scripts during that authorization method.

https://github.com/firebase/firebase-js-sdk/blob/master/packages/auth/src/rpchandler.js#L618

I had created the issue over on a firebase project

https://github.com/firebase/quickstart-js/issues/366

and I created a test case where the only differences from the example code provided to do the auth was serving it as application/xhtml+xml (and a few small changes to make the html file xml compliant). 

test.xhtml:
```
<?xml version="1.0" encoding="utf-8"?>
<html xmlns="http://www.w3.org/1999/xhtml"  lang="en">
<head>
<title>closure xhtml bug test</title>
<script src="closure-library/closure/goog/base.js"></script>
<script src="closuretest.js"></script>
</head>
<body onload="start()">
<h1>closure xhtml scriptelement test</h1>
<div id="test">testdiv</div>
</body>
</html>
```

test.html:
```
<!DOCTYPE html>
<html lang="en">
<head>
<title>closure xhtml bug test</title>
<script src="closure-library/closure/goog/base.js"></script>
<script src="closuretest.js"></script>
</head>
<body onload="start()">
<h1>closure xhtml scriptelement test</h1>
<div id="test">testdiv</div>
</body>
</html>
```

closuretest.js
```
// I had to delay this to let the script tag above this to finish loading first.
setTimeout(function(){
goog.require('goog.net.jsloader');
},100);

function start() {
         // likewise have to delay to ensure the require finishes
        setTimeout(function(){
            var url = goog.html.TrustedResourceUrl.fromConstant(goog.string.Const.from('loaded.js'));
            var result = goog.net.jsloader.safeLoad(url);
            console.log(result);
        },1500);
}
```

loaded.js:
```
console.log("I am a loaded script");
```

It seems as if the error is perhaps in how the scripts are added to the page? It seems like perhaps it's using doc.write - are we sure the element written is well-formed? If it wasn't, it would explain the error 

I got using firebase:

```
Failure: Argument is not a HTMLScriptElement (or a non-Element, non-Location mock); got: HTMLUnknownElement nexttick.js:45:38
```

Is there a way we can adjust the loading to continue to work with XHTML5 sites?

**Comment by ForbiddenEra** [comment]
Can we do something like this instead?

```
var el = document.createElement('div');
el.innerHTML = '<script></script>';
document.body.appendChild(el);
```

**Comment by ForbiddenEra** [comment]
Actually, it looks like in the else case that the script DOES create a script element without document.write..! So I guess the question is, perhaps, why do some cases use doc.write and what can be done to make this case not?

https://github.com/google/closure-library/blob/master/closure/goog/base.js#L3462

After reading Issue #844 I thought that adding ``var CLOSURE_DEFINES = {'goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING': true};`` would help, but doesn't seem to have changed the behaviour?

**Comment by ForbiddenEra** [comment]
It seems that defining it that way did not help. I also tried just doing ``goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING=true;``

with no luck.. how should I definte this?

**Comment by ForbiddenEra** [comment]
Okay, so I just edited line 342 of base.js to set ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING to true by default.. was really hoping that would be enough to fix the issue..but it's not:

http://localbuds.shaped.ca/closuretest/2/testnew.php

To be fair, enabling that flag and serving at as regular HTML also doesn't work..?

http://localbuds.shaped.ca/closuretest/2/testhtml.php

I'm tearing my hair out here trying to fix this..

**Comment by ForbiddenEra** [comment]
Well.. I just ran into something wierd..

So I thought I'd take the require and put it in a seperate script tag in the <head> before loading my script..

Then it (not on the xhtml version, but with ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING set to true) loaded my script.. but still gave an error?!

![Screenshot (63)](https://user-images.githubusercontent.com/7980490/64239262-2e279e80-cebd-11e9-94e0-6002a15b0ecb.png)

and if i refresh:

![Screenshot (64)](https://user-images.githubusercontent.com/7980490/64239346-544d3e80-cebd-11e9-9ff7-8ac1bc93d843.png)

Then I refreshed like 10 times and it loaded again as in the first image.. for it to say "i am a loaded script" then goog.net.jsloader has worked.

My brain isn't working anymore. I need help :(

**Comment by ForbiddenEra** [comment]
I can't even figure out what the heck the difference is.. like only SOMETIMES does it call back and get 'i am a loaded script' but most of the time it fails.. like I have been refreshing over and over and over and most times it fails but sometimes it works. I would swear I was on crack if I didn't have screenshots..

**Comment by ForbiddenEra** [comment]
AHA and same with the xHTML version it fired one time here:

![Screenshot (66)](https://user-images.githubusercontent.com/7980490/64240110-9d51c280-cebe-11e9-8ecd-64362479c986.png)

And then bam the same error that I get when trying to use popup auth with firebase:Failure: Argument is not a HTMLScriptElement (or a…ment, non-Location mock); got: HTMLUnknownElement

**Comment by ForbiddenEra** [comment]
I guess something's not always loading fast enough? setting a timeout in my code before trying to call safeload makes it "work" everytime..obviously by work I mean still throwing the Argument is not HTMLScriptElement error..

I went into the code and console.log'd the script variable as it's created through the code and just before the assertion check it console logs as: <SCRIPT type="text/javascript"></SCRIPT>

.. curiously, it's in CAPITALS!? which could (would?) break XHTML...BUT..closure is able to load it's own dependencies in this case, as you can see in the last screenshot, the XHTML pages header is populated with script tags. It's just my call to jsloader.safeload that seems to fail (and is the call that fails in firebase, which is the root of my whole issue)

**Comment by ForbiddenEra** [comment]
Okay, so I thought adding the timeout helped but it's only randomly working again.. I feel so dumb right now.. probably need a break..24h nonstop trying to figure out this bug..

Right..so the timeout needs to be on the require I guess..

Annnnnnnnd... BAM
it works if I set 

```
//  var script = goog.dom.createElement(goog.dom.TagName.SCRIPT);
  var script = goog.dom.createElement('script');
```

on line 165 of jsloader.js !!

so .. goog.dom.TagName definitions should be lowercase? at least for XHTML?!

**Comment by ForbiddenEra** [comment]
https://github.com/google/closure-library/blob/master/closure/goog/dom/tagname.js#L458

setting that to:
```
goog.dom.TagName.SCRIPT = new goog.dom.TagName('script');
```

resolves the error I believe.. Can I ask if there is a reason why the tag names are all uppercase? As they are just passed to the DOM to create tags and the like, it's not ever accessed by a coder so coding conventions don't matter (the variable name - goog.dom.TagName.SCRIPT can still be uppercase) but the actual tag should be created in lowercase, at least for XHTML.

Is there some sort of use-case or browser that requires tags to be upper-case? is this just a weird convention?

**Comment by ForbiddenEra** [comment]
Also, I still have to set

```
goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING =
    goog.define('goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING', true);
```

in line 320 of base.js - as mentioned above, I cannot seem to figure out how to assert this to true within my own code..?

**TL;DR:**

Issue: Using goog.net.jsloader with a page served with the content type application/xhtml+xml fails. Setting ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING helps the situation, but the injected tags are still in uppercase as defined in goog.dom.tagName, which causes an issue with pages again served with said content type as the tag names must be lowercase in XHTML.

Possible Solutions:
- Can we detect if we are using XHTML or in an environment that would require ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING to be set TRUE? If not, being able to set it myself is fine, although the only example I could find on how to set it did not work for me.

- If there's no other reason to have the tag names as uppercase, set them to lowercase in closer/goog/dom/tagname.js -- if there is an absolute required compatibility reason to have tag names as uppercase -- then is there a way to detect XHTML/XML and lowercase them? Otherwise is there any objection to setting them to all lowercase? As mentioned above, the variable names (for coding convention) could remain uppercase, just their values set to lowercase. I can't think of an instance where a browser would require uppercase tags.

**Comment by ForbiddenEra** [comment]
I have verified that the creation of an uppercase <SCRIPT> tag with jsloader is also the 100% the issue causing my popup not to work using Firebase with xHTML. Took a little while to verify / make my fix work because of the build system that is used I am not familiar with, in fact - with my minimum test case for testing with just closure, simply modifying tagname.js to have a lowercase script tag fixed it, along with enabling safe loading.. With firebase, that change didn't seem to help, neither did modifying the method to always return lowercase tagnames..but then I modified jsloader.js to have `` var script = goog.dom.createElement('script');`` on line 165 instead of referencing the tagname class and I was able to get my auth popup on firebase.

**Comment by ForbiddenEra** [comment]
If there's no objections to making the actual tag names lowercase or a solid reason for using uppercase, I would be happy to create a pull request.

Of course, we can keep the variable names uppercase for convention & compatibility, no code relying on this code should have to make any changes, while adding compatibility for XHTML.

**Comment by ForbiddenEra** [comment]
Maybe a lot of work to create a pull for the simple change I propose:

```
#:~/closure-library/closure/goog/dom$ diff tagname.js ~/tagname.js
37c37
<   this.tagName_ = tagName;
---
>   this.tagName_ = tagName.toLowerCase();
```

Simple and effective. Someone much more experienced with closure than I would have to say whether it could break anything but I would guess not. Perhaps the better change is to change every tag name in the file to be lowercase? Though the result should be equal I would think.

**Comment by ForbiddenEra** [comment]
So, apparently

```
// Closure Compiler unconditionally converts the following constants to their
// string value (goog.dom.TagName.A -> 'A'). These are the consequences:
// 1. Don't add any members or static members to goog.dom.TagName as they
//    couldn't be accessed after this optimization.
// 2. Keep the constant name and its string value the same:
//    goog.dom.TagName.X = new goog.dom.TagName('Y');
//    is converted to 'X', not 'Y'.
```

my fix works for uncompiled closure but not for compiled closure.. :(

**Comment by kjin** [comment]
Hi @ForbiddenEra -- I'm still going through your issue, but to start, it seems that you can set `ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING` to true as such ([docs](https://github.com/google/closure-library/blob/d78b31d31cef2cace4622919a3623a1abe9b8655/closure/goog/base.js#L86-L104)):

```html
<script>
  var CLOSURE_DEFINES = {
    'goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING': true
  };
</script>
<!--Include goog/base.js after for this to take effect.-->
```

**Comment by kjin** [comment]
I think we could fix this by changing our `goog.dom.createElement` implementation to always lowercase a tag name before creating an element:

```
// HTML
document.createElement('div').constructor.name; // 'HTMLDivElement'
document.createElement('div').tagName; // 'DIV'
document.createElement('DIV').constructor.name; // 'HTMLDivElement'
document.createElement('DIV').tagName; // 'DIV'
// XHTML
document.createElement('div').constructor.name; // 'HTMLDivElement'
document.createElement('div').tagName; // 'div'
document.createElement('DIV').constructor.name; // 'HTMLUnknownElement'
document.createElement('DIV').tagName; // 'div'
```

It doesn't seem to affect existing HTML cases, and fixes the XHTML case.

**Comment by ForbiddenEra** [comment]
That seems like it could be a suitable fix that the compiler wouldn't optimize out.

If you can point me to the file with that method I can give it a shot and provide a diff

Cheers,
Jai B.
IdealSound:Mobile Solutions for your Driving Experiencewww.IdealSound.ca
1 (587) 287-7420


----- Original Message -----

**Comment by kjin** [comment]
I'm thinking [here](https://github.com/google/closure-library/blob/d78b31d31cef2cace4622919a3623a1abe9b8655/closure/goog/dom/dom.js#L953), let us know how that works!

**Comment by ForbiddenEra** [comment]
Indeed my issue with setting the ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING was not the code itself but the fact I was calling it after loading base.js, for my closure-only test case at least.

closure-library/closure/goog/dom/dom.js line 953 becomes:
```
  return doc.createElement(String(name).toLowerCase());
```

And works fine for my un-compiled closure test-case. Now to test with Firebase...and..works compiled as well.

I'm not sure it's really worth the effort to create a pull request for such a small fix; I'm happy to have anyone associated with the project contribute it. Otherwise if needed, I can do a pull.

**Comment by kjin** [comment]
It's small enough that we can make a fix for it. Thanks!

**Comment by studiesma** [comment]
plz i need somebody to help me checking this [code](https://www.studies.ma/)

**Comment by nreid260** [comment]
Created Google internal issue b/140641626

**Comment by kjin** [comment]
This is fixed in b6a040835fc8b1e3c87341db287d98fced6cca9f and has already been released.

**Comment by ForbiddenEra** [comment]
Woot! Glad to hear it.. Thanks / you're welcome ? ;-)

**Comment by kjin** [comment]
We'll take both :)
